### PR TITLE
KOGITO-2895 & KOGITO-2896: Enabling PMML support on DMN Designer

### PR DESCRIPTION
**Pull request by yesamer**
@karreiro @jomarko Can you please review and test it?
Scope of these PRs are to fully enable the PMML usage inside the DMN Editor. 
The marshaller used to process the `*.pmml` is the one available on JS side, in kogito-tooling repo. For this reason, a js-bridge with retrieves the marshalled data was implemented.

Screencast from 08-11-2020 12_07_37 PM

Please note, currently only 2 kind of models are enabled: **ScoreCardModel** and **RegressionModel**. To enable **all** models, we need to merge this PR: https://github.com/kiegroup/kogito-tooling/pull/236

**Related PRs**
kie-wb-common: https://github.com/kiegroup/kie-wb-common/pull/3388
appformer: https://github.com/kiegroup/appformer/pull/1025
https://github.com/kiegroup/kogito-tooling/pull/248

Special credit to @karreiro and @manstis :)

https://issues.redhat.com/browse/KOGITO-2895
https://issues.redhat.com/browse/KOGITO-2896
https://issues.redhat.com/browse/BAPL-1689

### CODE REVIEW

File: appformer-kogito-bridge/src/test/java/org/appformer/kogito/bridge/client/pmmleditor/marshaller/PMMLEditorMarshallerServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
+package org.appformer.kogito.bridge.client.pmmleditor.marshaller;
```
Comment (karreiro) [inline appformer-kogito-bridge/src/test/java/org/appformer/kogito/bridge/client/pmmleditor/marshaller/PMMLEditorMarshallerServiceTest.java:1]: We're missing the copyright here :-)

File: appformer-kogito-bridge/src/test/java/org/appformer/kogito/bridge/client/pmmleditor/marshaller/PMMLEditorMarshallerServiceProducerTest.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
+package org.appformer.kogito.bridge.client.pmmleditor.marshaller;
```
Comment (karreiro) [inline appformer-kogito-bridge/src/test/java/org/appformer/kogito/bridge/client/pmmleditor/marshaller/PMMLEditorMarshallerServiceProducerTest.java:1]: We're missing the copyright here :-)

File: appformer-kogito-bridge/src/main/java/org/appformer/kogito/bridge/client/pmmleditor/marshaller/model/JSIUtils.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+                return toSet;
+            });
+        }
+        return toReturn;
+    }-*/;
+
+    public static<D> List<D> toList(final JsArrayLike<D> jsArrayLike) {
+        final List<D> toReturn = new ArrayList<>();
+        if (Objects.nonNull(jsArrayLike)) {
+            for (int i = 0; (i<jsArrayLike.getLength()); i ++) {
```
Comment (karreiro) [inline appformer-kogito-bridge/src/main/java/org/appformer/kogito/bridge/client/pmmleditor/marshaller/model/JSIUtils.java:61]: I believe this file is missing our auto-format.

Comment (yesamer) [reply appformer-kogito-bridge/src/main/java/org/appformer/kogito/bridge/client/pmmleditor/marshaller/model/JSIUtils.java]: @karreiro Done, thank you.

File: appformer-kogito-bridge/src/test/java/org/appformer/kogito/bridge/client/pmmleditor/marshaller/PMMLEditorMarshallerServiceProducerTest.java (hunk 2 above)
Comment (yesamer) [reply appformer-kogito-bridge/src/test/java/org/appformer/kogito/bridge/client/pmmleditor/marshaller/PMMLEditorMarshallerServiceProducerTest.java:16]: @karreiro Done, thank you.

File: appformer-kogito-bridge/src/test/java/org/appformer/kogito/bridge/client/pmmleditor/marshaller/PMMLEditorMarshallerServiceTest.java (hunk 1 above)
Comment (yesamer) [reply appformer-kogito-bridge/src/test/java/org/appformer/kogito/bridge/client/pmmleditor/marshaller/PMMLEditorMarshallerServiceTest.java:16]: @karreiro Done, thank you.

File: appformer-kogito-bridge/src/main/java/org/appformer/kogito/bridge/client/pmmleditor/marshaller/model/JSIUtils.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        // Utils class
+    }
+
+    /**
+     * Helper method to create a new, empty <code>JsArrayLike</code>
+     * @return
+     */
+    public static native <D> JsArrayLike<D> getNativeArray() /*-{
+        return [];
+    }-*/;
```
Comment (jomarko) [inline appformer-kogito-bridge/src/main/java/org/appformer/kogito/bridge/client/pmmleditor/marshaller/model/JSIUtils.java:37]: This is already in `org.kie.workbench.common.dmn.webapp.kogito.marshaller.mapper.JsUtils`

File: appformer-kogito-bridge/src/main/java/org/appformer/kogito/bridge/client/pmmleditor/marshaller/model/JSIUtils.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        var toReturn = [];
+        if (original != null) {
+            toReturn = original.map(function (arrayItem) {
+                var retrieved = arrayItem.value
+                var toSet = retrieved == null ? arrayItem : retrieved
+                return toSet;
+            });
+        }
+        return toReturn;
+    }-*/;
```
Comment (jomarko) [inline appformer-kogito-bridge/src/main/java/org/appformer/kogito/bridge/client/pmmleditor/marshaller/model/JSIUtils.java:56]: This is already in `org.kie.workbench.common.dmn.webapp.kogito.marshaller.mapper.JsUtils`

File: appformer-kogito-bridge/src/main/java/org/appformer/kogito/bridge/client/pmmleditor/marshaller/model/JSIUtils.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    public static<D> List<D> toList(final JsArrayLike<D> jsArrayLike) {
+        final List<D> toReturn = new ArrayList<>();
+        if (Objects.nonNull(jsArrayLike)) {
+            for (int i = 0; (i < jsArrayLike.getLength()); i++) {
+                final D toAdd = Js.uncheckedCast(jsArrayLike.getAt(i));
+                toReturn.add(toAdd);
+            }
+        }
+        return toReturn;
+    }
```
Comment (jomarko) [inline appformer-kogito-bridge/src/main/java/org/appformer/kogito/bridge/client/pmmleditor/marshaller/model/JSIUtils.java:67]: This is already in `org.kie.workbench.common.dmn.webapp.kogito.marshaller.mapper.JsUtils`

**Comment by yesamer** [comment]
> I think we shouldn't have both `org.kie.workbench.common.dmn.webapp.kogito.marshaller.mapper.JsUtils` and `org.appformer.kogito.bridge.client.pmmleditor.marshaller.model.JSIUtils`. I think we should keep juts the second one and extend it with content of first one. The first one should be deleted then.

@jomarko  `org.appformer.kogito.bridge.client.pmmleditor.marshaller.model.JSIUtils` is an auto generated file (by this mvn plugin https://github.com/kiegroup/gwt-jsonix-schema-compiler), therefore is not possible to modify it.

**Comment by jomarko** [comment]
Issue that needs to be fixed afterwards https://issues.redhat.com/browse/KOGITO-3152

**Comment by yesamer** [comment]
@jomarko I included the fix for https://issues.redhat.com/browse/KOGITO-3153 in this PR. Basically, the included model now holds only the filename and no more the relative path. Can you please retest? Thanks!

**Comment by yesamer** [comment]
@karreiro @jomarko @tiagobento Hi guys, it's crucial to include this feature in next release, so I kindly ask you to finalize the review and merge it by tomorrow. Thanks!

**Comment by yesamer** [comment]
jenkins execute compile downstream build

**Comment by karreiro** [comment]
@yesamer, @tiagobento is on PTO. So, I will assign @paulovmr as the final reviewer.

**Review by ederign — APPROVED** [review]

**Pull request merged**